### ISSUE TITLE
Error in babel-register 6.16.0

### ISSUE THREAD

**Issue by theverything** [babel/babel#4592]
After updating to babel register 6.16.0 I started getting the following error.

```
/myproject/node_modules/babel-core/lib/transformation/file/options/option-manager.js:319
        throw e;
        ^
SyntaxError:  (While processing preset: "/myproject/node_modules/babel-preset-es2015/lib/index.js")
```

.babelrc
```
{
  "presets": [
    ["es2015", { "modules": false }],
    "stage-1",
    "react"
  ],
  "plugins": [
    "transform-runtime"
  ],
  "env": {
    "development": {
      "plugins": ["react-hot-loader/babel", "transform-runtime"]
    },
    "test": {
      "presets": ["es2015", "stage-1", "react"],
    }
  },
  "compact": true
}
```

package.json
```js
{
  "dependencies": {
    ...
    "babel-core": "^6.3.26",
    "babel-loader": "^6.2.0",
    "babel-plugin-transform-runtime": "^6.12.0",
    "babel-polyfill": "^6.9.0",
    "babel-preset-es2015": "^6.13.0",
    "babel-preset-react": "^6.11.0",
    "babel-preset-stage-1": "^6.13.0",
    "babel-register": "^6.8.0",
    "babel-runtime": "^6.11.6",
    ...
  }
}
```

After resetting `babel-register` to 6.14.0 the error went away.

I'm thinking it might have something to do with this commit https://github.com/babel/babel/commit/71dff8c834191356bfc58d24f37e3601df96987e

| software         | version
| ---------------- | -------
| Babel            | 6.16.0
| node             | 6.6.0
| npm              | 3.10.3
| Operating System | OS X 10.11.6 (15G1004)

**Comment by hzoo** [comment]
Thanks, you'll just have to pin it for the moment.

**Comment by danielmcgrath** [comment]
We also started seeing a strange error:

```
node_modules/sinon/lib/sinon/util/core.js:153
            method.stackTrace = (new Error("Stack Trace for original")).stack;
                                                                       ^
NetworkError:
```

The release of babel 6.16.0 triggered this, and explicitly adding babel-register 6.14 to our package.json alleviates the issue.

**Comment by davidkaminsky** [comment]
Possibly related but seeing this strange error after updating to 6.16.0 in bluebird. Downgrading back to 6.14.0 fixed it.

node_modules/bluebird/js/release/debuggability.js:673
    var firstStackLines = firstLineError.stack.split("\n");
                                        ^
SyntaxError

**Comment by 3LOK** [comment]
+1 on this issue.

../node_modules/babel-register/node_modules/babel-core/lib/transformation/file/options/option-manager.js:319
        throw e;
        ^
SyntaxError:  (While processing preset: ".../node_modules/babel-preset-es2015/lib/index.js")

**Comment by didierfranc** [comment]
The same, downgrade to `babel-register to 6.14.0` to temporarily fix it.

**Comment by danez** [comment]
Anyone has a working example on how to reproduce this? 

If I install babel-register with presets and use it it works just fine. My bet would be we might have an issue with core 16 and presets <16 or the other way around, but couldn't prove that either.

**Comment by vjwilson** [comment]
@danez ,

I only see the error when I upgrade both `babel-core` and `babel-register`, and when I try to run Mocha tests, because I only use babel-register in my test setup helper, not my build process. (I use Webpack's babel-loader in the build.)

Here's the error I was getting:

        $ npm t
        
        > project-name@1.0.0 test /Users/vwilson/Projects/ex3-mocha-enzyme
        > mocha --reporter spec test/helpers/testSetup.js "test/**/*.test.js"
        
        
        /Users/vwilson/Projects/ex3-mocha-enzyme/node_modules/babel-core/lib/transformation/file/options/option-manager.js:319
                throw e;
                ^
        SyntaxError:  (While processing preset: "/Users/vwilson/Projects/ex3-mocha-enzyme/node_modules/babel-preset-es2015/lib/index.js")
        npm ERR! Test failed.  See above for more details.


I downgraded just babe-register, which seems to fix it, but I then downgraded babel-core, too, just to be safe.

**Comment by danez** [comment]
Can anyone with the problem post the output of `npm ls | grep babel`?

**Comment by 3LOK** [comment]
Hi,

@danez I created this example: https://github.com/3LOK/babel-test

Run npm install && npm test, should fail. Fix the version of babel-register and es2015 to 6.14 and it works.

**Comment by danez** [comment]
@3LOK aweesome thanks.

**Comment by danez** [comment]
In this example the node_modules are not ignored, and therefor also run thru babel and it seems this breaks the preset.

**Comment by 3LOK** [comment]
@danez Sorry! Will fix the example and change it to make sure the bug recreates (although it's weird that when fixing the version to 6.14 it works...)

**Comment by 3LOK** [comment]
@danez Ok, update the example to the minimal I can that doesn't work. Seems to me (in our specific case) to be related to the function ```initJsdom``` in spec/setup.js. 

Still investigating.

**Comment by danez** [comment]
One problem we found is that when node_modules are not ignored in babel-register then this error appears.

@3LOK The error with the updated example seems different. I'll check

**Comment by 3LOK** [comment]
@danez It's something specific about our overloading of some global variables for jsdom. 

Checking with the relevant developer where he got this overloading function from. If it's something he found online, maybe others use it as well.

**Comment by vjwilson** [comment]
@danez , @3LOK ,

It may be this snippet, which is recommended by the folks who make the enzyme library,
https://github.com/airbnb/enzyme/blob/master/docs/guides/jsdom.md

        /* setup.js */
        
        var jsdom = require('jsdom').jsdom;
        
        global.document = jsdom('');
        global.window = document.defaultView;
        Object.keys(document.defaultView).forEach((property) => {
          if (typeof global[property] === 'undefined') {
            global[property] = document.defaultView[property];
          }
        });
        
        global.navigator = {
          userAgent: 'node.js'
        };


I have added a call at the top of that:

        require('babel-register')();

Could that combination be causing the error in the new version?

**Comment by 3LOK** [comment]
@danez @vjwilson Yes it is - this is what's causing the problem for us. 

Suggested workarounds?

**Comment by vjwilson** [comment]
@danez ,

Can we pass a param to the function returned by the `require`.

Something like:

        require('babel-register')({ exclude: 'node_modules/' });

**Comment by danez** [comment]
By default babel-register already ignores node-modules.

**Comment by hzoo** [comment]
`"test": "mocha --compilers js:babel-register"` works fine so maybe ignore/exclude is still broken

**Comment by 3LOK** [comment]
@danez Ok, so my investigation is complete :)

In the enzyme solution, they suggest doing the following for every property of ```jsdom('').defaultView```:

```javascript
global[property] = document.defaultView[property];
```

One of the properties overloaded is ```XMLHttpRequest```. 

Adding the following solves the issue.

```javascript
    delete global['XMLHttpRequest'];
```

Can you please check why having a global.XMLHttpRequest screws up babel?

**Comment by danez** [comment]
The problem is already solved with #4606

**Comment by batjko** [comment]
I just spent all day trying to figure out why I'm getting this very error.
I eventually figured one of our developers must have introduced some strange dependency, since our test-related files were in fact identical to a previously working commit.

Only now, stumbling over this github issue, has it become clear. Jesus Bloody Christ! 

@3LOK 's suggestion worked for me as well. Thanks!

**Comment by danez** [comment]
This is now completely fixed with #4620
